python/cugraph/cugraph/gnn/pyg_extensions/data/cugraph_store.py

```diff
@@ -0,0 +1,737 @@
+            not accept PyG edge type tuples.
+
+        Returns
+        -------
+        The appropriate extracted subgraph.  Will extract the subgraph
+        if it has not already been extracted.
+
+        """
+        edge_types = tuple(sorted(edge_types))
+
+        if edge_types not in self.__subgraphs:
+            query = f'(_TYPE_=="{edge_types[0]}")'
+            for t in edge_types[1:]:
+                query += f' | (_TYPE_=="{t}")'
+            selection = self.__graph.select_edges(query)
+
+            # FIXME enforce int type
+            sg = self.__graph.extract_subgraph(
+                selection=selection,
+                default_edge_weight=1.0,
+                allow_multi_edges=True,
+                renumber_graph=True,
+                add_edge_data=False
+            )
+            self.__subgraphs[edge_types] = sg
+
+        return self.__subgraphs[edge_types]
+
+    def neighbor_sample(
+            self,
+            index,
+            num_neighbors,
+            replace,
+            directed,
+            edge_types):
+
+        if isinstance(num_neighbors, dict):
+            # FIXME support variable num neighbors per edge type
+            num_neighbors = list(num_neighbors.values())[0]
+
+        # FIXME eventually get uniform neighbor sample to accept longs
+        if self.backend == 'torch' and not index.is_cuda:
+            index = index.cuda()
+        index = cupy.from_dlpack(index.__dlpack__())
+
+        # FIXME resolve the directed/undirected issue
+        G = self._subgraph([et[1] for et in edge_types])
+
+        index = cudf.Series(index)
+        if self.is_mg:
+            sampling_results = cugraph.dask.uniform_neighbor_sample(
+                G,
+                index,
+                # conversion required by cugraph api
+                list(num_neighbors),
+                replace
+            )
+        else:
+            sampling_results = cugraph.uniform_neighbor_sample(
+                G,
+                index,
+                # conversion required by cugraph api
+                list(num_neighbors),
+                replace
+            )
+
+        concat_fn = dask_cudf.concat if self.is_mg else cudf.concat
+
+        nodes_of_interest = concat_fn(
+            [sampling_results.destinations, sampling_results.sources]
+        ).unique()
+
+        noi = self.__graph.get_vertex_data(
+            nodes_of_interest.compute() if self.is_mg else nodes_of_interest,
+            columns=[self.__graph.vertex_col_name, self.__graph.type_col_name]
+        )
+
+        noi_types = noi[self.__graph.type_col_name].unique()
+        if len(noi_types) > 1:
+            noi = noi.groupby(self.__graph.type_col_name)
+        else:
+            class FakeGroupBy:
+                def __init__(self, df):
+                    self.df = df
+
+                def get_group(self, g):
+                    return self.df
+            noi = FakeGroupBy(noi)
```

Comment:
The performance of `groupby` in general is so bad that I'm taking all of this code out in a future PR that uses an efficient solution for both homogeneous and heterogeneous graphs.  So for this PR I'm going to drop the optimized version since it's all going away eventually.